packages/ra-core/src/dataProvider/withLifecycleCallbacks.spec.ts

```diff
@@ -0,0 +1,854 @@
+    describe('afterSave', () => {
+        it('should alter the create result data', async () => {
+            const base = {
+                create: jest.fn((resource, { data }) =>
+                    Promise.resolve({ data: { id: 1, ...data } })
+                ),
+            };
+            const dataProvider = withLifecycleCallbacks(base, [
+                {
+                    resource: 'posts',
+                    afterSave: jest.fn(record =>
+                        Promise.resolve({
+                            ...record,
+                            foo: 'bar',
+                        })
+                    ),
+                },
+            ]);
+
+            const result = await dataProvider.create('posts', {
+                data: { title: 'foo' },
+            });
+
+            expect(result).toEqual({
+                data: { id: 1, title: 'foo', foo: 'bar' },
+            });
+        });
+        it('should alter the update result data', async () => {
+            const base = {
+                update: jest.fn((resource, { id, data }) =>
+                    Promise.resolve({ data: { id, ...data } })
+                ),
+            };
+            const dataProvider = withLifecycleCallbacks(base, [
+                {
+                    resource: 'posts',
+                    afterSave: jest.fn(record =>
+                        Promise.resolve({
+                            ...record,
+                            foo: 'bar',
+                        })
+                    ),
+                },
+            ]);
+
+            const result = await dataProvider.update('posts', {
+                id: 1,
+                data: { title: 'foo' },
+                previousData: { title: 'bar' },
+            });
+
+            expect(result).toEqual({
+                data: { id: 1, title: 'foo', foo: 'bar' },
+            });
+        });
+        it('should be called on the updateMany', async () => {
+            const base = {
+                getMany: jest.fn((resource, { ids }) =>
+                    Promise.resolve({
+                        data: ids.map(id => ({ id })),
+                    })
+                ),
+                updateMany: jest.fn((resource, { ids, data }) =>
+                    Promise.resolve({ data: ids })
+                ),
+            };
+            const resourceCallback = {
+                resource: 'posts',
+                afterSave: jest.fn(record =>
+                    Promise.resolve({
+                        ...record,
+                        foo: 'bar',
+                    })
+                ),
+            };
+            const dataProvider = withLifecycleCallbacks(base, [
+                resourceCallback,
+            ]);
+
+            const result = await dataProvider.updateMany('posts', {
+                ids: [1, 2],
+                data: { title: 'foo' },
+            });
+
+            expect(result).toEqual({
+                data: [1, 2],
+            });
```

Comment:
I don't understand this test. I'm even worried that the callback is indeed called but does nothing to the resulting value.